### PR TITLE
Don't let a slot do anything special if it isn't in a shadow tree

### LINKED ISSUE #447: Shadow: Specify when `slotchange` fires
Safari, Chrome and ShadyDOM currently exhibit different behavior when `slotchange` does and does not fire. As far as I understand the spec, it is not precisely spec’d, meaning all these current behaviors are technically spec compliant, but can be very frustrating for developers.

[Demo 1](https://jsbin.com/liveko/9/edit?html,output): Slots are created in the constructor
[Demo 2](https://jsbin.com/puzilim/3/edit?html,output): Slots are creatred in `connectedCallback`

**Which browser fires `slotchange` under what circumstances?**

| | Slots in constructor, parser creates element  | Slots in constructor, element gets upgraded | Slots in `connectedCallback`, parser creates element | Slots in `connectedCallback`, element gets upgraded |
|-|----|---|---|---|
|Chrome| Fire | No fire | Fire | No Fire |
|Safari| Fire | No fire | No Fire | No Fire |
|ShadyDOM| Fire | Fire | Fire | Fire |

(ShadyDOM was tested using Firefox).

While **my main concern is consistency across browsers**, I like ShadyDOM’s behavior best in terms of DX as it is consistent with how `attributeChangeCallback` in that it fires upon initialization. This would allow developers to have their update logic in the `slotchange` handler and they wouldn’t need to manually detect silently slotted elements in `connectedCallback()`.

/cc @hayatoito

### PR DESCRIPTION
**Pull request by hayatoito**
Fixes #447.

***
[Preview](https://s3.amazonaws.com/pr-preview/hayatoito/dom/slotchange-when-inserted.html) | [Diff](https://s3.amazonaws.com/pr-preview/whatwg/dom/d9cd30d...hayatoito:e1db083.html)

### CODE REVIEW

**Comment by hayatoito** [comment]
@surma is taking care of the PR for the tests:  https://github.com/w3c/web-platform-tests/pull/5705

**Comment by hayatoito** [comment]
It looks we have to update the existing test in WPT.
I am working on that at https://github.com/w3c/web-platform-tests/pull/5954.

**Comment by annevk** [comment]
This change LGTM. We only need bugs against Chrome and WebKit I suspect, since nobody else is far enough along thus far.

**Comment by hayatoito** [comment]
Yeah, regarding Chrome, I'll take care of it.

**Comment by hayatoito** [comment]
I have updated the PR so that it can reflect the conclusion in #447.
@domenic , could you have a chance to review this?
cc: @annevk

File: dom.bs [hunk 1]
```diff
@@ -1738,32 +1741,23 @@ for a given <a>slot</a> <var>slot</var>, run these steps:</p>
...
+<p>To <dfn noexport>assign slotables</dfn> for a <a>slot</a> <var>slot</var>, run these steps:
 
 <ol>
  <li><p>Let <var>slotables</var> be the result of <a>finding slotables</a> for <var>slot</var>.
 
- <li><p>If <var>suppress signaling flag</var> is unset, and <var>slotables</var> and
- <var>slot</var>'s <a for=slot>assigned nodes</a> are not identical, then run
- <a>signal a slot change</a> for <var>slot</var>.
+ <li><p>If <var>slotables</var> and <var>slot</var>'s <a for=slot>assigned
+ nodes</a> are not identical, then run <a>signal a slot change</a> for <var>slot</var>.
```
Comment (domenic) [inline dom.bs]: I notice that this instance of "signal a slot change" has fewer checks before it than the other two. Is that correct? (I don't understand the algorithms that well so I am just pattern-matching in the hopes of finding something useful :)

Comment (domenic) [inline dom.bs]: Nit: annevk prefers tags all on one line in his specs, so no linebreak in the middle of an `<a>stuff stuff</a>`. So the whole `<a for=slot>assigned nodes</a>` should get pushed to the next line.

Comment (hayatoito) [inline dom.bs]: Yes, that is correct and intentional.
- For assigned nodes' changes, the condition became relaxed.
- For other cases (aka fallback contents), the condition become more tight.

Comment (hayatoito) [inline dom.bs]: Thanks. Let me put it in the same line.

**Comment by hayatoito** [comment]
@domenic Thank you for the review.
I have updated the PR, by squashing the commits.

I am now working on updating WPT. I'll ping @rniwa in PR for WPT too.

File: dom.bs [hunk 2]
```diff
@@ -1719,7 +1722,8 @@ for a given <a>slot</a> <var>slot</var>, run these steps:</p>
 
   <ol>
    <li>
-    <p>If <var>node</var> is a <a>slot</a>, then:
+    <p>If <var>node</var> is a <a>slot</a> whose <a for=tree>root</a> is a <a for=/>shadow root</a>,
```
Comment (hayatoito) [inline dom.bs]: I also updated this line, which I forgot to fix in the previous PR.

**Comment by hayatoito** [comment]
@rniwa, could you have a chance to take a look?

PR for WPT Tests is here: https://github.com/w3c/web-platform-tests/pull/5954

**Comment by hayatoito** [comment]
FYI. I implemented this new behavior in Blink.

**Comment by hayatoito** [comment]
@rniwa, friendly ping. I am assuming that this change sounds good to all of us.

**Comment by rniwa** [comment]
The intention seems fine but did you check that we don't end up signaling the same slot multiple times per a single DOM change especially nested ones?

I haven't gotten around to fix WebKit's implementation but I intend to do it in near future.

**Comment by hayatoito** [comment]
> The intention seems fine but did you check that we don't end up signaling the same slot multiple times per a single DOM change especially nested ones?

That shouldn't happen, as far as my understanding is correct. A slotchange is signaled at most once per a slot per a single DOM change for any tree structures.

**Comment by rniwa** [comment]
I understand that's the intention but it's really hard for me to tell if that's what's gonna happen with this change or not.

In general, I find it really hard to read these spec diffs and figure out what's actually happening. It would be ideal if we could look at the spec with the change applied. Is there a way to do that?

**Comment by hayatoito** [comment]
It looks this DOM repository adds a *Preview* or *Diff* link in the first comment in PR automatically.
I guess such a nice feature was recently introduced.

Though Diff link looks ugly for unknown reasons (Hmm, I remember that it showed a clean diff when I updated this PR), I think Preview link still works.

**Comment by hayatoito** [comment]
@annevk  Thanks. I have rebased on master and force push.

**Comment by annevk** [comment]
Great!

@rniwa you can review changes in a rendered way starting here (I think there's nothing before this point): https://s3.amazonaws.com/pr-preview/whatwg/dom/d9cd30d...hayatoito:32e6669.html#find-flattened-slotables.

**Comment by rniwa** [comment]
That diff link was very useful! The change looks good to me.

r=me.

File: dom.bs [hunk 3]
```diff
@@ -1738,32 +1742,23 @@ for a given <a>slot</a> <var>slot</var>, run these steps:</p>
...
+<p>To <dfn noexport>assign slotables</dfn> for a <a>slot</a> <var>slot</var>, run these steps:
 
 <ol>
  <li><p>Let <var>slotables</var> be the result of <a>finding slotables</a> for <var>slot</var>.
 
- <li><p>If <var>suppress signaling flag</var> is unset, and <var>slotables</var> and
- <var>slot</var>'s <a for=slot>assigned nodes</a> are not identical, then run
- <a>signal a slot change</a> for <var>slot</var>.
+ <li><p>If <var>slotables</var> and <var>slot</var>'s <a for=slot>assigned nodes</a> are not
+identical, then run <a>signal a slot change</a> for <var>slot</var>.
```
Comment (annevk) [inline dom.bs]: Could you add a space before identical? (I would do it, but I cannot modify this branch for reasons unclear.)

Comment (hayatoito) [inline dom.bs]: My bad. Done. I've updated the PR.

**Comment by annevk** [comment]
Due to jetlag I forgot to complete the usual checklist:

* Tests haven't landed yet. @hayatoito will take care of that per https://github.com/w3c/web-platform-tests/pull/5705#issuecomment-305745169.
* Browser bugs haven't been filed. @rniwa and @hayatoito, I suspect you'll take care of that for your respective engines? Nobody else implements this yet so I'm guessing nothing is needed for them.

**Comment by hayatoito** [comment]
Thanks for merging.
Regarding tests, I have updated existing tests for slotshange to reflect this change in another PR: https://github.com/w3c/web-platform-tests/pull/5954
I'll review https://github.com/w3c/web-platform-tests/pull/5705#issuecomment-305745169 too.

Regarding implementation, I could take care of Blink. I think @rniwa would take care of WebKit.

**Comment by annevk** [comment]
Thanks, I'm going to assume all is in order then.

**Pull request merged**